Initial single and concurrent requests

heaths opened the pull request:
Resolves #10691

heaths committed:
Initial single and concurrent requests

Resolves #10691

pakrym commented on sdk/keyvault/Azure.Security.KeyVault.Secrets/tests/ChallengeBasedAuthenticationPolicyTests.cs:
```diff
@@ -0,0 +1,233 @@
+
+        [TestCase(10, 0, 0)]
+        [TestCase(10, 0, 500)]
+        public async Task MultipleRequests(int numberOfRequests, int minDelay, int maxDelay)
```
Does this consistently repro the bug?

heaths replied:
It should since it basically replicates the repro app I wrote, but I can effectively back out my fix and make sure.

pakrym replied:
Yes, let's see if it reproes the issue.

heaths replied:
After a few tweaks, I was able to get this to fail about 2/5 of the time on my laptop with the older runtime code.

pakrym commented on sdk/keyvault/Azure.Security.KeyVault.Secrets/tests/ChallengeBasedAuthenticationPolicyTests.cs:
```diff
@@ -0,0 +1,233 @@
+            };
+
+            using AzureEventSourceListener logger = AzureEventSourceListener.CreateTraceLogger(EventLevel.Verbose);
+            SecretClient client = new SecretClient(s_vaultUri, new MockCredential(transport), options);
```
Do we need to go though SecretClient here? Can we write a test that only uses policy in question?

heaths replied:
It's possible, but why write more code? I'd rather keep the tests as close to production code as possible, and this is the most straight-forward way. Unless you feel strongly, I'd rather keep this as-is.

pakrym replied:
I would expect that testing only that policy would involve less code that going through the `SecretClient`.

Usually I try to keep tests that verify race conditions very targeted because they tend to be pretty sensitive and the more moving parts there are the harder it gets to keep the tests consistently reproducing the issue.

heaths committed:
More tests and perf improvements

heaths committed:
Remove Identity tests from solution

They require a special environment and stopped working 100% a while ago, and impact the successful completion of all tests here.

heaths commented:
I'm going to push this to the upstream remote as well and run live tests before we merge into the hotfix branch on the upstream remote.

pakrym commented on sdk/keyvault/Azure.Security.KeyVault.Shared/src/ChallengeBasedAuthenticationPolicy.cs:
```diff
@@ -133,22 +138,23 @@ public override bool Equals(object obj)
+                // This assumes that Authority Scopes are always non-null and Scopes has a length of one.
                 // This is guaranteed by the way the AuthenticationChallenge cache is constructed.
-                return Scopes[0].GetHashCode();
+                return Authority.GetHashCode() ^ Scopes[0].GetHashCode();
```
use hashcodebuilder.

pakrym commented on sdk/keyvault/Azure.Security.KeyVault.Secrets/tests/ChallengeBasedAuthenticationPolicyTests.cs:
```diff
@@ -76,23 +80,59 @@ public async Task MultipleRequests(int numberOfRequests, int minDelay, int maxDe
+            SecretClientOptions options = new SecretClientOptions
+            {
+                Transport = transport,
+                Diagnostics =
```
Remove debugging related stuff.

heaths replied:
Why? It's useful when debugging next time. All secret information is fake by design. I see no value in removing it.

pakrym replied:
What value is there in keeping it? Should we add it to every client we create in case we need to debug later?

pakrym replied:
To me it's similar to leaving Console.WriteLine everywhere just in case that never get cleaned up.

heaths replied:
That's oversimplifying it. Concurrency issues are difficult enough to diagnose, which this whole issue is about from the start. But the multiple hops against multiple endpoints makes this even harder. Unless you feel very strongly about this, I prefer to leave it in.

For tests in other public and private repos, yes, I tend to leave stuff in that makes it easier to diagnose test failures. Why remove it if you need it again? It's there when needed, and only happens when you debug tests and only with fake data in this particular case (no PII leaks).

pakrym replied:
Chances are this test would never fail again or if it does it would be completely different issue that would require more/different logging.

I feel strongly about consistency and don't want to encourage people leaving diagnostics enabled on per test basis. I'm open to having a common solution/convention for logging in tests but am against one offs.

heaths replied:
I'm a fan of consistency as well, but I feel it's not always required. I will remove it from all but the main concurrent method, and I'll comment it out so it's not left running in tests until needed. It should be easy to figure out how to restore what I did in my TDD changes to repro and track down the issue. Yes, hopefully they don't fail again and that's why we have tests, but if they do it's more or less ready to go.

pakrym commented on sdk/keyvault/Azure.Security.KeyVault.Shared/src/ChallengeBasedAuthenticationPolicy.cs:
```diff
@@ -225,27 +232,39 @@ private static AuthenticationChallenge ParseBearerChallengeHeaderValue(string ch
                             {
-                                if (string.Equals(key, "scope", StringComparison.OrdinalIgnoreCase))
+                                // Ordered by current likelihood.
+                                if (string.Equals(key, "authorization", StringComparison.OrdinalIgnoreCase))
```
Are we adding features?

heaths replied:
It actually wasn't correct before. I talked this over with @schaabs.

pakrym replied:
Whould we add a test for it?

heaths replied:
It is being tested with the third test.

heaths committed:
Resolve PR feedback
